File: 496. Next Greater Element I/Solution.py
class Solution:
    def nextGreaterElement(self, nums1: List[int], nums2: List[int]) -> List[int]:
        mapping = {}
        stack = []
        result = []
        
        stack.append(nums2[0])
        for i in range(1, len(nums2)):
            while stack and stack[-1] < nums2[i]:
                mapping[stack[-1]] = nums2[i]
                stack.pop()
            stack.append(nums2[i])
        
        for i in range(len(nums1)):
            if nums1[i] in mapping:
                result.append(mapping[nums1[i]])
            else:
                result.append(-1)
        
        return result


# Enter stack reversely
class Solution:
    def nextGreaterElement(self, nums1: List[int], nums2: List[int]) -> List[int]:
        mapping = {}
        stack = []
        result = []
        
        for i in range((len(nums2) - 1), -1, -1):
            while stack and stack[-1] < nums2[i]:
                stack.pop()
            if stack == []:
                mapping[nums2[i]] = -1
            else:
                mapping[nums2[i]] = stack[-1]
            stack.append(nums2[i])
        
        for el in nums1:
            result.append(mapping[el])
        
        return result

File: 496. Next Greater Element I/Solution.java
class Solution {
    public int[] nextGreaterElement(int[] nums1, int[] nums2) {
        int len1 = nums1.length;
        int len2 = nums2.length;
        int output[] = new int[len1];
        HashMap<Integer, Integer> map = new HashMap<>();
        Stack<Integer> stack = new Stack<>();
        
        for (int i = len2 - 1; i >= 0; i--) {
            while (!stack.isEmpty() && nums2[i] > stack.peek()) {
                stack.pop();
            }
            if (!stack.isEmpty()) {map.put(nums2[i], stack.peek());}
            stack.push(nums2[i]);
        }
        

        for (int i = 0; i < len1; i++) {
            output[i] = map.getOrDefault(nums1[i], -1);
        }
        return output;
    }
}